File: apps/ccusage/src/adapter/goose/paths.ts
import { existsSync, statSync } from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import process from 'node:process';
import { createFixture } from 'fs-fixture';

export const GOOSE_PATH_ROOT_ENV = 'GOOSE_PATH_ROOT';
export const GOOSE_DB_FILE_NAME = 'sessions.db';

function isFileSyncSafe(filePath: string): boolean {
	try {
		return statSync(filePath).isFile();
	} catch {
		return false;
	}
}

function getGooseDefaultDbCandidates(): string[] {
	const home = os.homedir();
	return [
		path.join(home, '.local/share/goose/sessions', GOOSE_DB_FILE_NAME),
		path.join(home, 'Library/Application Support/goose/sessions', GOOSE_DB_FILE_NAME),
		path.join(home, '.local/share/Block/goose/sessions', GOOSE_DB_FILE_NAME),
	];
}

export function getGooseDbPathFromRoot(root: string): string {
	return path.join(root, 'data/sessions', GOOSE_DB_FILE_NAME);
}

export function getGooseDbPaths(): string[] {
	const envRoot = process.env[GOOSE_PATH_ROOT_ENV]?.trim();
	const candidates =
		envRoot == null || envRoot === ''
			? getGooseDefaultDbCandidates()
			: [getGooseDbPathFromRoot(envRoot)];
	return Array.from(new Set(candidates.map((candidate) => path.resolve(candidate)))).filter(
		isFileSyncSafe,
	);
}

export function hasGooseDatabase(): boolean {
	return getGooseDbPaths().some((dbPath) => existsSync(dbPath));
}

if (import.meta.vitest != null) {
	describe('Goose path discovery', () => {
		afterEach(() => {
			vi.unstubAllEnvs();
		});

		it('uses GOOSE_PATH_ROOT/data/sessions/sessions.db when present', async () => {
			await using fixture = await createFixture({
				data: {
					sessions: {
						[GOOSE_DB_FILE_NAME]: '',
					},
				},
			});
			vi.stubEnv(GOOSE_PATH_ROOT_ENV, fixture.path);

			expect(getGooseDbPaths()).toEqual([
				path.resolve(fixture.getPath('data/sessions', GOOSE_DB_FILE_NAME)),
			]);
		});

		it('returns no paths for a missing GOOSE_PATH_ROOT database', async () => {
			await using fixture = await createFixture({});
			vi.stubEnv(GOOSE_PATH_ROOT_ENV, fixture.path);

			expect(getGooseDbPaths()).toEqual([]);
		});
	});
}

File: rust/crates/ccusage/src/adapter/goose/paths.rs
use std::{collections::HashSet, env, path::PathBuf};

use crate::Result;

pub(super) const GOOSE_PATH_ROOT_ENV: &str = "GOOSE_PATH_ROOT";
pub(super) const GOOSE_DB_FILE_NAME: &str = "sessions.db";

pub(super) fn goose_db_paths() -> Result<Vec<PathBuf>> {
    let candidates = if let Ok(root) = env::var(GOOSE_PATH_ROOT_ENV) {
        let root = root.trim();
        if root.is_empty() {
            default_goose_db_candidates()?
        } else {
            vec![
                PathBuf::from(root)
                    .join("data")
                    .join("sessions")
                    .join(GOOSE_DB_FILE_NAME),
            ]
        }
    } else {
        default_goose_db_candidates()?
    };

    let mut paths = Vec::new();
    let mut seen = HashSet::new();
    for path in candidates {
        let path = path.canonicalize().unwrap_or(path);
        if path.is_file() && seen.insert(path.clone()) {
            paths.push(path);
        }
    }
    Ok(paths)
}

fn default_goose_db_candidates() -> Result<Vec<PathBuf>> {
    let home =
        crate::home::home_dir().ok_or_else(|| crate::cli_error("home directory is not set"))?;
    Ok(vec![
        home.join(".local/share/goose/sessions")
            .join(GOOSE_DB_FILE_NAME),
        home.join("Library/Application Support/goose/sessions")
            .join(GOOSE_DB_FILE_NAME),
        home.join(".local/share/Block/goose/sessions")
            .join(GOOSE_DB_FILE_NAME),
    ])
}

#[cfg(test)]
mod tests {
    use std::{env, path::Path, sync::Mutex};

    use super::*;
    use ccusage_test_support::fs_fixture;

    static ENV_LOCK: Mutex<()> = Mutex::new(());

    struct EnvDirGuard;

    impl EnvDirGuard {
        fn set(dir: &Path) -> Self {
            unsafe { env::set_var(GOOSE_PATH_ROOT_ENV, dir) };
            Self
        }
    }

    impl Drop for EnvDirGuard {
        fn drop(&mut self) {
            unsafe { env::remove_var(GOOSE_PATH_ROOT_ENV) };
        }
    }

    #[test]
    fn discovers_goose_path_root_database() {
        let _guard = ENV_LOCK.lock().unwrap();
        let fixture = fs_fixture!({
            "data/sessions/sessions.db": "",
        });
        let _cleanup = EnvDirGuard::set(fixture.root());

        let paths = goose_db_paths().unwrap();

        assert_eq!(paths.len(), 1);
        assert!(paths[0].ends_with(Path::new("data/sessions/sessions.db")));
    }
}
